src/librustdoc/passes/collect_intra_doc_links.rs

```diff
@@ -1089,10 +1107,61 @@ impl LinkCollector<'_, '_> {
         ori_link: &str,
         link_range: Option<Range<usize>>,
     ) -> Option<(Res, Option<String>)> {
+        let lookup_key = (disambiguator.map(Disambiguator::ns), ori_link.to_owned());
+
+        // Try to look up both the result and the corresponding side channel value
+        if let Some((res, side_channel)) = self.visited_links.get(&lookup_key) {
```

Comment:
`ori_link` is not the right key here. a) it already includes disambiguator, so you don't need to hash that again, and b) it doesn't include information about `crate::` and `Self::` which was replaced in `resolve_link`.

I'm not sure what the correct hash would be ... until `Self::` is handled in one place (https://github.com/rust-lang/rust/pull/76467) `current_item` still makes a difference. Maybe something like this?

- item
- path_str
- current_node
- extra_fragment

I would refactor that into a separate struct so it's not too unwieldy. Now that `current_node` is being hashed you should be able to get rid of the cache changes for new modules too.

Please also add some test cases for `crate::` and `Self::`. To trigger the current issues you'll need `crate::foo` on the same link in different modules and `Self::x` on `impl`entations of two different types.

Reply:
Is hashing `item` a good idea? That would mean duplicate errors reported if other items also have wrong links. Which is technically fine, but would be the result of re-visiting already known links.

Reply:
Ah hmm I need to double check how item is being used. Maybe it's only for diagnostics?

Reply:
It's an input of `report_diagnostic`.

Reply:
It's also a bit confusing since there is `item` which is I think the item the docstring is on, and `current_item` which I've been ignoring so far.

Reply:
Ok yes it's only for diagnostics. Then `(path_str, current_node, extra_fragment)` should work.

Reply:
> current_item which I've been ignoring so far 

Ah right, yeah this is going to mess up resolution of `Self`. I need to fix up https://github.com/rust-lang/rust/pull/76467 before you can make this change I think.

Reply:
If I use `base_node, path_str.to_owned(), extra_fragment`, some tests always fail. Up until now, it was prim_conflict, but now I get this, which is is core:

```
error: incompatible link kind for `write`
   --> library/core/src/fmt/mod.rs:166:32
    |
166 |     /// Glue for usage of the [`write!`] macro with implementors of this trait.
    |                                ^^^^^^^^ help: to link to the function, add parentheses: ``write()``
    |
    = note: `-D broken-intra-doc-links` implied by `-D warnings`
    = note: this link resolved to a function, which is not a macro
```

I think I need to dig a bit deeper :)

Reply:
Ah sorry, you need to include the disambiguator too in the cache key.

Reply:
It's only `orig_link` that includes the disambiguator, `path_str` doesn't.